pkg/operator/controller/controller_dns_configmap.go

```diff
@@ -212,20 +294,76 @@ func corefileChanged(current, expected *corev1.ConfigMap) (bool, *corev1.ConfigM
 	return true, updated
 }
 
-func coreDNSResolver(upstream operatorv1.Upstream) (string, error) {
+// caBundleRevisionMap generates a map of ca bundle configmaps with their resource versions
+// to be used in Corefile as the path of ca bundles. Resource version is appended to enable automatic
+// reload of Corefile when there is a change in the source ca bundle configmap (e.g. cert rotation).
+func caBundleRevisionMap(client client.Client, dns *operatorv1.DNS) (map[string]string, error) {
+	caBundleRevisions := make(map[string]string)
+	transportConfig := dns.Spec.UpstreamResolvers.TransportConfig
+	if transportConfig.Transport == operatorv1.TLSTransport {
+		if transportConfig.TLS != nil && transportConfig.TLS.CABundle.Name != "" {
+			sourceName := types.NamespacedName{
+				Namespace: GlobalUserSpecifiedConfigNamespace,
+				Name:      transportConfig.TLS.CABundle.Name,
+			}
+			cm := &corev1.ConfigMap{}
+			if err := client.Get(context.TODO(), sourceName, cm); err != nil {
```

Comment:
Done.

Reply:
The problem is that, when we generate the caBundleRevisionMap, the sync has not happened yet. So, there is no configmaps on `openshift-dns`. Now that I moved the function to controller.go, would it be ok that it's using the configmaps from `openshift-config`?

Reply:
The recommended pattern for OpenShift operators is that the user creates configmaps and secrets in the `openshift-config` namespace, the operator synchronizes resources from that namespace to the operator's namespace, and then the operator uses the resources from its namespace.  Often this is implemented with a discrete controller that performs the synchronization: controller A synchronizes from `openshift-config` to `openshift-foo`, and controller B reads from `openshift-foo` to do whatever work the operator does that requires the resources.  This pattern isolates these discrete steps, simplifies the operator's logic, and enables it to maintain full control over access, validation, referential integrity, updates, etc.  (Unfortunately we didn't follow this pattern from the start with some operators and resources, but it is the recommended pattern.)  

Here, the logic is a little entangled:  The same controller is synchronizing the configmaps between `openshift-config` and `openshift-dns` and then reading the original configmaps from `openshift-config` to configure the CoreDNS daemonset and Corefile configmap.  This makes the logic more difficult to reason about and verify that nothing breaks if, for example, the user deletes or update a configmap that the operator is using.  

Using separate controllers isn't strictly necessary here.  However, it would still be preferable to keep synchronization separate from usage.  Would it be feasible to have `ensureDNS` (A) call `ensureCABundleConfigMaps` as a discrete step before `ensureDNSDaemonSet` and `ensureDNSConfigMap` to synchronize the configmaps to `openshift-dns`, and then (B) call these latter two methods, which would use strictly only the configmaps from `openshift-dns`?  

If what I'm suggesting would require a significant rewrite, we can leave it as a follow-up.

Reply:
I've changed the implementation as you described, and, luckily, it didn't require significant rewrite.